kernel/sched.c

```diff
@@ -34,12 +34,34 @@ static void update_cache(int preempt_ok);
 static void halt_thread(struct k_thread *thread, uint8_t new_state);
 static void add_to_waitq_locked(struct k_thread *thread, _wait_q_t *wait_q);
 
-
 BUILD_ASSERT(CONFIG_NUM_COOP_PRIORITIES >= CONFIG_NUM_METAIRQ_PRIORITIES,
 	     "You need to provide at least as many CONFIG_NUM_COOP_PRIORITIES as "
 	     "CONFIG_NUM_METAIRQ_PRIORITIES as Meta IRQs are just a special class of cooperative "
 	     "threads.");
 
+#ifdef CONFIG_SCHED_DEADLINE
+/*
+ * Compare two deadlines
+ *
+ * > 0 -> d1 occurs before d2
+ * = 0 -> d1 and d2 are the same deadline
+ * < 0 -> d1 occurs after d2
+ */
+int32_t z_sched_deadline_cmp(uint32_t d1, uint32_t d2)
+{
+	/*
+	 * An earlier deadline means a higher effective priority. Since all
+	 * deadlines are assumed to live within the same "half" of the 32-bit
+	 * modulus space, then all deadline differences in the queue are
+	 * guaranteed to be be (2's complement) non-negative.
+	 *
+	 * Using unsigned types and casting to signed isn't perfect, but at
+	 * least reduces this from UB on overflow to impdef.
+	 */
+	return (int32_t) (d2 - d1);
```

Comment:
At the risk of being too naive with C integers once more:

```C
return d1 < d2 ? d2 - d1 : - (d1 - d2);
```

Doesn't that avoid `impdef`?

I disassembled the output of clang -O2 (version 15) and on X86_64 the code generated is strictly identical to `return d2-d1;`